PSReadLine/ConsoleLib.cs

```diff
@@ -97,7 +97,32 @@ public Encoding OutputEncoding
             set { try { Console.OutputEncoding = value; } catch { } }
         }
 
-        public ConsoleKeyInfo ReadKey()                  => Console.ReadKey(true);
+        private void WaitForKeyAvailable()
+        {
+            // On Unix platforms input echo is on by default. If we wait for KeyAvailable without
+            // disabling it, all input will be echoed.
+            UnixConsoleEcho.InputEcho.Disable();
+            try
+            {
+                while (!Console.KeyAvailable)
+                {
+                    System.Threading.Thread.Sleep(50);
```

Comment:
I don't know about Linux, but on Windows, a sleep for less than 200ms or so (I forget exact details) prevents the CPU from going into low power mode because the kernel assumes the process is busy.

So I'm a bit concerned this might cause increased drain of the battery.

Is there some other way where we can block in the kernel instead of a busy loop?

Reply:
I'm not having luck finding a way to reliably block until a key is available.  But a possible solution would be to set the sleep to 200ms until the first key is pressed, then switch to `SpinWait.SpinUntil` with a timeout of a couple seconds.  Then switch back to the 200ms sleep if the timeout is hit.  Do you have any concerns with that approach?

Reply:
That also sounds less than ideal.

If you want to unblock - maybe we could agree on an escape sequence that you generate and PSReadLine ignores, other than to break out of the Console.ReadKey call of course.

Reply:
Hmm, that's an interesting thought.  Early on when I was first looking into the issue for editor services I was experimenting with emitting the terminal information query escape sequence to break out of `ReadKey`.  That worked but wasn't reliable because of different possible return sequences.  If it's always the same and always ignored, that may work perfectly.  I'll do some testing.

Reply:
I think we'd need an invalid escape sequence so there's 0 chance the tty will interpret it. Alternatively, maybe we should use an [invalid Unicode character](http://www.unicode.org/faq/private_use.html).

Reply:
I'm remembering now that I did try to go this route initially, but I couldn't find a way to actually write to stdin directly.  At least not in a way that would release `Console.ReadKey`.  The closest I've come is query escape sequences written to stdout that emit back to stdin.  None of those that I'm aware of allow you to specify what they return.

Still digging, I would definitely prefer this method as well.  If you have any suggestions or hints, they would be greatly appreciated.

Reply:
I only have hints, no direct experience.

On Windows, one can call `CreateFile` with the filename `CONIN$` and you will have a file handle to the keyboard. I've never tried writing to it, so it's possible it might not work.

I know less about Linux.  Maybe one of the files in `/dev` will work in a similar manner, e.g. `/dev/stdin` or `/dev/tty` - but I'm totally guessing here. You might look at how the tool `expect` is implemented, though it's possible that tool just redirects `stdin` when it creates the process.

Another option might be to just start powershell/pwsh with redirected input. On Linux, .Net converts the VT escape sequences generated by modified keys (e.g. `Alt-a`) to what PSReadLine expects. On Windows 10, thanks to @parkovski - PSReadLine can also do that mapping, but you need to set the console mode flag [ENABLE_VIRTUAL_TERMINAL_INPUT](https://docs.microsoft.com/en-us/windows/console/setconsolemode).

With a redirected stdin - the language service would own sending every key to PSReadLine, so it's possible you wouldn't need any of these changes. It does mean not supporting <Win10, or alternatively generating VT escape sequences for <Win10, and that might require some PSReadLine changes.

Hopefully this helps somehow - I'd really like to avoid shipping multiple platform dependent native libraries (and thanks for pointing out that x86 isn't supported, I scanned the PowerShell releases too quickly.)

Reply:
At this point I'm reasonably confident it isn't possible.  I've tried PInvoking `write(2)` and writing directly to the file descriptor 0 (stdin).  That comes out as output, which from what I've read appears to be by design (or at least to be expected).  Just to see if I could get *something* to work I tried inserting into various private buffers stored by `ConsolePal` and `StdInReader`, but it's already waiting on native code by the time `ReadKey` is blocking.

I really like the idea of redirecting and controlling all input, but that'd be a hefty undertaking for a (hopefully) temporary work around.  Ideally, this issue would be solved within corefx.

That said, did you have any concerns about the cancellation token changes from the first commit?  That's the only change required for PSReadLine to work within Windows.  I already have the logic to fall back to the current readline implementation for PowerShell 3/4.  If the busy loop and native libraries are enough of a blocker, I can remove the entire second commit from the PR and only support PSReadLine on Windows for the time being.

Reply:
I'd like to look into this. There's still an issue with cursor position escapes when pwsh.exe is run from WSL, and the only way I could get close to making it work is by manually buffering the console input. I think if I can get that working, I'd have a version of `ReadKey` that blocks on an event and not the actual console.

Reply:
That'd be amazing @parkovski.  If you can figure that out you'd be my hero.

Some more information, [here's the PR](https://github.com/PowerShell/PowerShellEditorServices/pull/592) for editor services where I go into a little bit more detail on the cause.  And here's [the issue](https://github.com/PowerShell/vscode-powershell/issues/987) where I go through some of the things I tried. Good luck :)

Reply:
@lzybkr I'm not sure how much better it is, but I do have another thought.  If a better implementation isn't possible/feasible, maybe PSReadLine could surface an API to customize `ReadKey`.  Editor Services will need to ship these native binaries and use the busy loop either way, if we could give PSReadLine a `ReadKey` delegate then we could drop the second commit and still support Linux/Mac.  That wouldn't fix the event logic outside of PSES though.

Reply:
It's not public, but there is already an interface to mock `Console.ReadKey` for the tests. The tests just change the mock via reflection, so you could do the same as an experiment.

That said, `stdin` feels like it is that "delegate". Ensuring `stdin` works for nearly everything helps in scenarios beyond PSES. (I think a few bindings are impossible on Windows with VT escape sequences but are possible with `Console.ReadKey`.)

Reply:
I found a way to block until input is received.  I can p/invoke [poll(2)](http://man7.org/linux/man-pages/man2/poll.2.html), also from libc.  That has an option to block indefinitely until input is received.  It works, but it also triggers on `Console.CursorLeft` which spills some of the escape sequence into `Console.ReadKey`.  Granted this is the exact problem corefx was trying to solve by locking.

If I can figure out a good way to filter that out, coupled with replacing the native libs with p/invokes, that should hopefully solve all issues.

Reply:
The problem with filtering that out is whoever called `CursorLeft` will get an invalid result, which can make line editing glitchy, unless you can put it back in the input stream before anything else is read. It might work to make a new tty and pass the original input through a filter, but to make that reliable I think you'd have to predict the terminal's responses, which seems like not a good idea. I'm leaning toward the only way to solve this is by patching CoreCLR, unfortunately.

Reply:
@parkovski Yeah I'm operating under the assumption I can find a way to determine if poll was triggered because of `CursorLeft` without actually consuming the input.  If I can't, then I agree it's isn't feasible.

Reply:
Unfortunately I ran into more issues.  Changing the terminal settings via p/invoke isn't working as expected.  I think there is an additional setting or two that is getting changed because new lines aren't triggering input unless `CTRL` is held.

On top of that, it looks like the `termios` stuct can be very different between architectures.  Some have more members, some flags have different values, and the control characters array can vary in size.  For some of those, getting them wrong results in a pretty nasty crash. That said, my experience with p/invoke (and Linux in general) is pretty limited, so [here's the code](https://gist.github.com/SeeminglyScience/18a38347625c47c11017908d4dc5874d) I was testing with.  It's very possible I missed something, but I don't think p/invoke will be reliable in this instance.

For Editor Services I'm going to move forward with the native binaries (or `stty -echo`) and the long/short busy loop because the Unix debugging experience is completely broken without it.  But  PSReadLine in VSCode may need be Windows only until the issue is fixed in corefx.

@lzybkr How would you like to proceed?  Assuming you don't have any issues with the first commit, should I remove the second or open a new PR?